google-api-client/src/main/java/com/google/api/client/googleapis/auth/oauth2/DefaultCredentialProvider.java

```diff
@@ -256,19 +258,45 @@ private GoogleCredential getCredentialUsingWellKnownFile(
     }
   }
 
-
-  private boolean runningOnAppEngine() {
-    if (getEnvEquals("GAE_ENV", "standard")) {
+  private boolean useGAEStandardAPI() {
+    if (getEnvEquals("GAE_ENV", "standard") && getEnvEquals("GAE_RUNTIME", "java7")) {
       return true;
     }
-    if (getEnvEquals("GAE_RUNTIME", "java7")) {
-      return true;
+
+    Class<?> systemPropertyClass;
+    try {
+      systemPropertyClass = forName("com.google.appengine.api.utils.SystemProperty");
+    } catch (ClassNotFoundException expected) {
+      // SystemProperty will always be present on App Engine.
+      return false;
     }
-    if (getEnvEquals("GAE_RUNTIME", "java8")) {
-      return true;
+    Exception cause = null;
+    Field environmentField;
```

Comment:
Maybe add a comment about what you're doing here, since it's a bit difficult to decipher what happens with the reflection.

Reply:
Done